hugolib/content_index.go

```diff
@@ -0,0 +1,86 @@
+// Copyright 2016 The Hugo Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hugolib
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"path/filepath"
+	"time"
+
+	"github.com/spf13/hugo/helpers"
+	jww "github.com/spf13/jwalterweatherman"
+	"github.com/spf13/viper"
+)
+
+type index struct {
+	Pages []*entry
+}
+
+type entry struct {
+	Title      string      `json:"title"`
+	Content    string      `json:"content"`
+	Permalink  string      `json:"permalink"`
+	Tags       interface{} `json:"tags"`
+	Categories interface{} `json:"categories"`
+	Date       time.Time   `json:"date"`
+	LastMod    time.Time   `json:"lastmod"`
+}
+
+// createJSONIndex creates a JSON-based index of
+// all content files and it's metadata
+func createJSONIndex(pages Pages) {
+	rawIndex := new(index)
+
+	for _, p := range pages {
+		if !p.IsDraft() && !p.IsFuture() && p.IsRenderable() {
+			entry := &entry{
+				Title:      p.Title,
+				Content:    p.plain,
+				Permalink:  getPermalink(p),
+				Tags:       p.Params["tags"],
+				Categories: p.Params["categories"],
+				Date:       p.Date,
+				LastMod:    p.Lastmod,
+			}
+
+			rawIndex.Pages = append(rawIndex.Pages, entry)
+		}
+	}
+
+	jsonIndex, err := json.Marshal(rawIndex.Pages)
```

Comment:
While I see the benefit of using json.Marshal to generate the JSON, this approach is also quite rigid defining the explicit fields that will be rendered/generated. 

What if instead we used a similar approach to how the sitemap works using a template instead to generate the output. It would be much more flexible.